models/equipment.js

```diff
@@ -0,0 +1,227 @@
+      ],
+    );
+    const category = (await query(
+      'SELECT * FROM "categories" WHERE "id" = ($1) LIMIT 1',
+      [
+        categoryId,
+      ],
+    )).rows[0];
+    return category;
+  });
+
+  return Promise.all(promisedCategories).then(categories => {
+    return { ...createdEquipment, categories };
+  });
+};
+
+exports.find = async id => {
+  const equipment = (await query(
+    'SELECT * FROM "equipments" WHERE "id" = $1 LIMIT 1',
+    [
+      id,
+    ],
+  )).rows[0];
+  return equipment;
+};
+
+exports.findByName = async name => {
+  const category = (await query(
+    'SELECT * FROM "equipments" WHERE "name" = $1 LIMIT 1',
+    [
+      name,
+    ],
+  )).rows[0];
+  return category;
+};
+
+exports.update = async newProperties => {
+  const oldProps = await this.find(newProperties.id);
+  const properties = { ...oldProps, ...newProperties };
+
+  const errors = await validate(properties);
+  if (errors) {
+    return { errors };
+  }
+
+  const updatedEquipment = (await query(
+    `UPDATE "equipments" SET
+    "name"=$1,
+    "totalForCheckout"=$2 WHERE id=$3 RETURNING *`,
+    [
+      properties.name,
+      properties.totalForCheckout,
+      properties.id,
+    ],
+  )).rows[0];
+
+  const oldCategoryIds = (await query(
+    `SELECT
+      categories.id,
+      categories.name
+    FROM categorizations, categories
+    WHERE categorizations."equipmentId" = ($1)
+    AND categorizations."categoryId" = categories.id;`,
+    [
+      properties.id,
+    ],
+  )).rows.map(category => category.id);
+
+
+  const newCategoryIds =
+    (properties.categoryIds === undefined ? []
+    : typeof properties.categoryIds === 'string' ? [ properties.categoryIds ]
+    : [ ...properties.categoryIds ]).map(stringId => Number(stringId));
+
+  const categorizationsToDelete = oldCategoryIds.filter(oldId => !newCategoryIds.includes(oldId));
+  const categorizationsToCreate = newCategoryIds.filter(newId => !oldCategoryIds.includes(newId));
+
+  categorizationsToCreate.map(async categoryId => {
+    await query(
+      `INSERT INTO "categorizations"(
+        "categoryId",
+        "equipmentId"
+      ) values ($1, $2) returning *`,
+      [
+        categoryId,
+        properties.id,
+      ],
+    );
+  });
+
+  await categorizationsToDelete.map(async categoryId => {
+    await query(
+      `DELETE FROM "categorizations"
+        WHERE "categoryId" = $1
+        AND "equipmentId" = $2
+      returning *`,
+      [
+        categoryId,
+        properties.id,
+      ],
+    );
+  });
+
+  const categories = (await query(
+    `SELECT
+      categories.id,
+      categories.name
+    FROM categorizations, categories
+    WHERE categorizations."equipmentId" = ($1)
+    AND categorizations."categoryId" = categories.id;`,
+    [
+      properties.id,
+    ],
+  )).rows;
+
+  return { ...updatedEquipment, categories };
+};
+
+async function validate(properties) {
+  const errors = [];
+  if (properties.name === '') {
```

Comment:
This isn't an api, so not of upmost concern... but what if a create request comes in without a properties.name or properties.totalForCheckout? the database should reject it, but the server might fall over. awkward.

Reply:
Done